### ISSUE TITLE
What is a directory stream?

### ISSUE THREAD

**Issue by johnwfinigan** [johnwfinigan/CSCI2467-Fall2015#7]
A student asks "what is a directory stream" - here is my answer, in rough terms that are close enough for what we are doing.

A "stream" is UNIX jargon for a certain type of IO interface, and here we are talking about directory streams. You can think of a directory stream as an "open directory object", that is, it provides a connection between your program and a directory, such that you can read the directory's entries. Here is what `man 3 opendir` says:

```
   The  opendir()  function  opens a directory stream corresponding to the
   directory name, and returns a pointer to  the  directory  stream.   The
   stream is positioned at the first entry in the directory.
```

When you call opendir successfully, it returns a pointer to a `DIR` struct. This is your "handle" to the open directory stream. The `DIR` struct itself is an "opaque struct", meaning that you do not need to be concerned with its contents, which are implementation specific anyway -- they're private to the operating system's developers. All you need to worry about is the the pointer itself, which is your connection to the open directory stream.

Once you have a `DIR *` directory stream pointer, you can pass it to `readdir_r`, which has this prototype from `man 3 readdir` 

```
   int readdir_r(DIR *dirp, struct dirent *entry, struct dirent **result);
```

Its first argument is a `DIR *`, that is, a directory stream. Then, each time you call `readdir_r`  , it reads one entry from the directory the stream is connected to, assuming it is successful.

So, in summary, a directory stream is a connection to an open directory, and in your program you can get one by calling `opendir` and storing its return value in a `DIR *` pointer, which is your reference to the stream as long as it is open.
